agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1082 @@
+		if err := c.UpdateRoots(roots); err != nil {
+			return err
+		}
+		args.QueryOptions.MinQueryIndex = nextIndexVal(args.QueryOptions.MinQueryIndex, roots.QueryMeta.Index)
+		return nil
+	}, func(err error) {
+		c.logger.Error("CA root replication failed, will retry",
+			"routine", secondaryCARootWatchRoutineName,
+			"error", err,
+		)
+	})
+
+	return nil
+}
+
+// UpdateRoots updates the cached roots from the primary and regenerates the intermediate
+// certificate if necessary.
+func (c *CAManager) UpdateRoots(roots structs.IndexedCARoots) error {
+	// Update the state first to claim the 'lock'.
+	if err := c.setState(CAStateReconfig, true); err != nil {
+		return err
+	}
+	defer c.setState(CAStateReady, false)
+
+	// Update the cached primary roots now that the lock is held.
+	if err := c.setPrimaryRoots(roots); err != nil {
+		return err
+	}
+
+	// Check to see if the primary has been upgraded in case we're waiting to switch to
+	// secondary mode.
+	provider, _ := c.getCAProvider()
+	if provider == nil {
+		// this happens when leadership is being revoked and this go routine will be stopped
+		return nil
+	}
+	if !c.configuredSecondaryCA() {
+		versionOk, primaryFound := ServersInDCMeetMinimumVersion(c.srv, c.srv.config.PrimaryDatacenter, minMultiDCConnectVersion)
+		if !primaryFound {
+			return fmt.Errorf("Primary datacenter is unreachable - deferring secondary CA initialization")
+		}
+
+		if versionOk {
+			if err := c.initializeSecondaryProvider(provider, roots); err != nil {
+				return fmt.Errorf("Failed to initialize secondary CA provider: %v", err)
+			}
+		}
+	}
+
+	// Run the secondary CA init routine to see if we need to request a new
+	// intermediate.
+	if c.configuredSecondaryCA() {
+		if err := c.initializeSecondaryCA(provider, nil); err != nil {
+			return fmt.Errorf("Failed to initialize the secondary CA: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// initializeSecondaryProvider configures the given provider for a secondary, non-root datacenter.
+// It is being called while holding the stateLock in order to update actingSecondaryCA, which means
+// it must never take that lock itself or call anything that does.
+func (c *CAManager) initializeSecondaryProvider(provider ca.Provider, roots structs.IndexedCARoots) error {
+	if roots.TrustDomain == "" {
+		return fmt.Errorf("trust domain from primary datacenter is not initialized")
+	}
+
+	clusterID := strings.Split(roots.TrustDomain, ".")[0]
+	_, conf, err := c.srv.fsm.State().CAConfig(nil)
+	if err != nil {
+		return err
+	}
+
+	pCfg := ca.ProviderConfig{
+		ClusterID:  clusterID,
+		Datacenter: c.srv.config.Datacenter,
+		IsPrimary:  false,
+		RawConfig:  conf.Config,
+		State:      conf.State,
+	}
+	if err := provider.Configure(pCfg); err != nil {
+		return fmt.Errorf("error configuring provider: %v", err)
+	}
+
+	return c.setSecondaryCA()
+}
+
+// setSecondaryCA sets the flag for acting as a secondary CA to true.
+func (c *CAManager) setSecondaryCA() error {
```

Comment:
Should this also pick up a `validate bool` arg like the `setState` function did?

Reply:
I don't think it needs one yet, since there's no place that wants to bypass the validation currently